backend/de.metas.material/planning/src/main/java/de/metas/material/planning/pporder/impl/PPRoutingRepository.java

```diff
@@ -304,6 +333,24 @@ public void changeRouting(@NonNull final PPRoutingChangeRequest changeRequest)
 				.list();
 	}
 
+	private List<I_PP_WF_Node_Product> retrieveProducts(@NonNull final I_AD_Workflow routingRecord)
```

Comment:
i think you can simply the logic from here by sending `Set<PPRoutingActivityId> activityIds` as parameter.

In case the Set is empty, u shall return empty right away.

In this query you can use something like 
```java
.addInArrayFilter(I_PP_WF_Node_Product.COLUMNNAME_AD_WF_Node_ID, activityIds)
```